src/dune_engine/build_system.ml

```diff
@@ -30,20 +30,100 @@ end = struct
 end
 
 module Error = Build_config.Error
-module Handler = Build_config.Handler
+
+module Progress = struct
+  type t =
+    { number_of_rules_discovered : int
+    ; number_of_rules_executed : int
+    }
+
+  let equal { number_of_rules_discovered; number_of_rules_executed } t =
+    Int.equal number_of_rules_discovered t.number_of_rules_discovered
+    && Int.equal number_of_rules_executed t.number_of_rules_executed
+
+  let complete t = t.number_of_rules_executed
+
+  let remaining t = t.number_of_rules_discovered - t.number_of_rules_executed
+
+  let is_determined { number_of_rules_discovered; number_of_rules_executed } =
```

Comment:
Not really related to this PR but I have trouble understanding what `is_determined` means. I think `number_of_rules_executed <> 0` implies `number_of_rules_discovered <> 0`, so this expression can be simplified. If that's true, can we just expose `number_of_rules_discovered`, whose meaning is much clearer than `is_determined`?

Reply:
(It's fine to just turn this into a `CR-someday` comment if you'd rather not change anything here.)

Reply:
This function is unused. I removed it.